Feature/auth and roles

HellHunterMax opened the pull request:
THIS WILL (probably) NOT BE MERGED WITH DEV THIS IS JUST FOR REFERENCE.
I WILL ADD MORE INFO AND DOCUMENTATION TO THIS TONIGHT!

chrisK00 commented on Server/JokesOnYou.Web.Api/Controllers/AuthController.cs:
```diff
@@ -7,16 +10,53 @@
+            }*/
+        }
+
+        [AllowAnonymous]
```
should work without the allow anonymous since its default

HellHunterMax replied:
Ah yes you are right I forgot to add the [Authorize] attribute. I want the Writer of the code to be explicit in his commands so when something should be accessible to everyone it should be [AllowAnonymous]. for the case of adding a new endpoint as well.

I was also reading about DefaultIdentity There is defaults to authorize everything. that got me mixed up as well.

But either way would work, removing it or adding authorize.

chrisK00 commented on Server/JokesOnYou.Web.Api/Controllers/AuthController.cs:
```diff
@@ -7,16 +10,53 @@
+                var user = await _authService.LoginAsync(userLoginDto);
+                return Ok(user);
+            }
+            catch (UserLoginException ex)
```
we should setup the global exception handler, so good idea to save possible exceptions.

HellHunterMax replied:
Yes we should I need to read up on global exception handlers ;) though this is not an unhandled exception so this should not give an issue and I think is good practice.

chrisK00 commented on Server/JokesOnYou.Web.Api/Controllers/AuthController.cs:
```diff
@@ -7,16 +10,53 @@
-        public Task<ActionResult> Register([FromBody] UserRegisterDTO userRegisterDto)
+        [AllowAnonymous]
+        [HttpPost("login")]
+        public async Task<IActionResult> Login(UserLoginDTO userLoginDto)
```
specify that a user will be sent back

HellHunterMax replied:
How ?

chrisK00 replied:
Change the iactionresult to ActionResult<User. Remove the Ok() that wraps the user. This will give us type safety and an Ok response is default when sending an item back i think

HellHunterMax replied:
![image](https://user-images.githubusercontent.com/39444801/113650703-f41e1000-9690-11eb-9dd9-5cb070189076.png)
that didnt work but this did:
![image](https://user-images.githubusercontent.com/39444801/113650860-3ba49c00-9691-11eb-99a3-773eccc15f39.png)

chrisK00 commented on Server/JokesOnYou.Web.Api/Controllers/AuthController.cs:
```diff
@@ -7,16 +10,53 @@
+            }
+            catch (UserLoginException ex)
+            {
+                return BadRequest(ex.Message);
```
return Unauthorized

HellHunterMax replied:
Oké so 401 it is ;)

`**400 Bad Request**`
The server cannot or will not process the request due to an apparent client error (e.g., malformed request syntax, size too large, invalid request message framing, or deceptive request routing).
`**401 Unauthorized**`
Similar to 403 Forbidden, but specifically for use when authentication is required and has failed or has not yet been provided.

It seems to be the standard.

source: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_client_errors

chrisK00 commented on Server/JokesOnYou.Web.Api/Extensions/ServiceExtensions.cs:
```diff
@@ -18,8 +21,29 @@ public static IServiceCollection ConfigureAppServices(this IServiceCollection se
 
-            services.AddIdentity<User, IdentityRole>().AddEntityFrameworkStores<DataContext>();
+            services.AddIdentity<User, IdentityRole>(
+                options => { options.SignIn.RequireConfirmedAccount = false; options.User.RequireUniqueEmail = true; }
```
sign in requireconfirmed account is fale by default i believe. Might wanna check if same applies to requniqmail

HellHunterMax replied:
Property                         | Description                                           | Default
RequireConfirmedEmail | Requires a confirmed email to sign in. | false


Property                   | Description                                                    | Default
RequireUniqueEmail | Requires each user to have a unique email. | false

chrisK00 commented on Server/JokesOnYou.Web.Api/Repositories/UserRepository.cs:
```diff
@@ -16,6 +18,26 @@ public UserRepository(UserManager<User> userManager)
+                var message = string.Empty;
+                foreach (var error in userCreationResult.Errors)
+                {
+                    message += error.Description+ System.Environment.NewLine;
```
This is some nice validation, maybe wanna use stringbuilder for performance

chrisK00 commented on Server/JokesOnYou.Web.Api/Repositories/UserRepository.cs:
```diff
@@ -27,6 +49,11 @@ public async Task<User> GetUserAsync(string id)
             return user;
```
shouldnt return user if he is deleted

HellHunterMax replied:
ah this is a weird one in Github. there are some lines missing above here. You can expand to see the missing lines.
DeleteUserAsync does not return a user :P
But you made me look at Looking up users and found a null error ;)

chrisK00 commented on Server/JokesOnYou.Web.Api/Services/AuthService.cs:
```diff
@@ -1,22 +1,60 @@
+                //TODO need to assign user a role!
+                UserReplyDTO userReplyDTO = new UserReplyDTO() { Email = user.Email, Id = user.Id, UserName = user.UserName, Role = "Registered" };
+
+                string token = _tokenService.GetToken(user);
```
for some more decoupling, maybe our token creator should just recieve what it needs, like the roles and username

HellHunterMax replied:
I'm not so sure about this one.
We could but would mean more work sometimes.
```
Every time we want to add anything to the GetToken we have to change this method and the GetToken.
If we add more places where GetToken gets called, they all have to change their parameters when GetToken changes.
```
```
Now if we add stuff to the Token we don't need to change this nor the parameters.
```
So i think this is better. But please do convince me!

chrisK00 commented on Server/JokesOnYou.Web.Api/Services/JwtTokenService.cs:
```diff
@@ -1,17 +1,62 @@
+            var tokenDescriptor = new SecurityTokenDescriptor
+            {
+                Subject = new ClaimsIdentity(new[] { new Claim(ClaimTypes.Name, user.Id), new Claim(ClaimTypes.Role, "Registered") }),
+                Expires = DateTime.UtcNow.AddDays(7),
```
we can set this to a day

HellHunterMax replied:
yes we could do anything with it :D

chrisK00 reviewed (commented):
just a couple of thoughts. Well done overall! Looks great.

HellHunterMax committed:
removed AddJsonOptions;
There is no loop in the app that gets used by json right now so this is not needed now.
might implement again when a loop arrises.

Commit:
created appexception

HellHunterMax committed:
Changed Return type of Login, renamed variable;

Changed Return type for Type Safty.
Renamed the var for better readabillity.

HellHunterMax committed:
Added UserServiceTests;

HellHunterMax committed:
Merge from dev and added AppException
